File: Python/utils/csv_reader.py
from typing import List, Type, TypeVar
from datetime import datetime
import csv

from models.attendee import Attendee
from models.event import Event
from models.registration import Registration


T = TypeVar("T", bound=Event | Attendee | Registration)

EVENTS_CSV_FILE = "Problem/events.csv"
ATTENDEES_CSV_FILE = "Problem/attendees.csv"
REGISTRATIONS_CSV_FILE = "Problem/registrations.csv"


class CsvReader:
    def __init__(self):
        self._events: List[Event] = []
        self._attendees: List[Attendee] = []
        self._registrations: List[Registration] = []

    @property
    def events(self) -> List[Event]:
        return self._events

    @property
    def attendees(self) -> List[Attendee]:
        return self._attendees

    @property
    def registrations(self) -> List[Registration]:
        return self._registrations

    def _read_data(self, csv_file: str, data_class: Type[T]) -> List[T]:
        with open(csv_file, "r", newline='') as f:
            reader = csv.reader(f)
            next(reader)  # skip the header
            result = []
            for row in reader:
                if issubclass(data_class, Attendee):
                    data = Attendee(
                        row[0].strip(),
                        row[1].strip(),
                    )
                elif issubclass(data_class, Event):
                    data = Event(
                        row[0].strip(),
                        datetime.strptime(
                            row[1].strip(),
                            "%m/%d/%Y %H:%M").date(),
                        row[2].strip(),
                    )
                else:
                    data = Registration(
                        int(row[0].strip()),
                        int(row[1].strip()),
                        datetime.strptime(
                            row[2].strip(),
                            "%m/%d/%Y %H:%M").date(),
                    )
                result.append(data)
            return result

    def read_events(self) -> None:
        self._events = self._read_data(EVENTS_CSV_FILE, Event)

    def read_attendees(self) -> None:
        self._attendees = self._read_data(ATTENDEES_CSV_FILE, Attendee)

    def read_registrations(self) -> None:
        self._registrations = self._read_data(
            REGISTRATIONS_CSV_FILE, Registration)

File: Go/utils/csv_reader.go
package utils

import (
	"encoding/csv"
	"fmt"
	"os"
	"strconv"
	"time"

	"github.com/siddhant-vij/Event-Management-System/models"
)

const (
	EVENTS_CSV_FILE        = "Problem/events.csv"
	ATTENDEES_CSV_FILE     = "Problem/attendees.csv"
	REGISTRATIONS_CSV_FILE = "Problem/registrations.csv"
)

type CsvReader struct {
	events        []*models.Event
	attendees     []*models.Attendee
	registrations []*models.Registration
}

func NewCsvReader() *CsvReader {
	return &CsvReader{
		events:        make([]*models.Event, 0),
		attendees:     make([]*models.Attendee, 0),
		registrations: make([]*models.Registration, 0),
	}
}

func (c *CsvReader) GetEvents() []*models.Event {
	return c.events
}

func (c *CsvReader) GetAttendees() []*models.Attendee {
	return c.attendees
}

func (c *CsvReader) GetRegistrations() []*models.Registration {
	return c.registrations
}

func (c *CsvReader) ReadDateFromCSV(tableName string) {
	switch tableName {
	case "events":
		c.readFromCSV(EVENTS_CSV_FILE, c.readEventsFromCSV)
	case "attendees":
		c.readFromCSV(ATTENDEES_CSV_FILE, c.readAttendeesFromCSV)
	case "registrations":
		c.readFromCSV(REGISTRATIONS_CSV_FILE, c.readRegistrationsFromCSV)
	default:
		fmt.Printf("invalid table name: %s\n", tableName)
	}
}

func (c *CsvReader) readFromCSV(csvFilePath string, reader func(br *csv.Reader) error) {
	file, err := os.Open(csvFilePath)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	err = reader(csv.NewReader(file))
	if err != nil {
		panic(err)
	}
}

func (c *CsvReader) readEventsFromCSV(br *csv.Reader) error {
	records, err := br.ReadAll()
	if err != nil {
		return err
	}

	for _, record := range records {
		if len(record) == 0 || record[0] == "event_name" {
			continue
		}
		eventName := record[0]
		eventDate, err := ParseDate(record[1])
		if err != nil {
			return err
		}
		eventLocation := record[2]

		event := models.NewEvent(eventName, eventLocation, eventDate)

		c.events = append(c.events, event)
	}

	return nil
}

func (c *CsvReader) readAttendeesFromCSV(br *csv.Reader) error {
	records, err := br.ReadAll()
	if err != nil {
		return err
	}

	for _, record := range records {
		if len(record) == 0 || record[0] == "attendee_name" {
			continue
		}
		attendeeName := record[0]
		attendeeEmail := record[1]

		attendee := models.NewAttendee(attendeeName, attendeeEmail)

		c.attendees = append(c.attendees, attendee)
	}

	return nil
}

func (c *CsvReader) readRegistrationsFromCSV(br *csv.Reader) error {
	records, err := br.ReadAll()
	if err != nil {
		return err
	}

	for _, record := range records {
		if len(record) == 0 || record[0] == "event_id" {
			continue
		}
		event_id, err := strconv.Atoi(record[0])
		if err != nil {
			return err
		}
		attendee_id, err := strconv.Atoi(record[1])
		if err != nil {
			return err
		}
		registration_date, err := ParseDate(record[2])
		if err != nil {
			return err
		}

		registration := models.NewRegistration(event_id, attendee_id, registration_date)

		c.registrations = append(c.registrations, registration)
	}

	return nil
}

func ParseDate(dateString string) (time.Time, error) {
	layouts := []string{
		"01/02/2006 15:04",
		"01/02/2006 5:04",
		"01/2/2006 15:04",
		"01/2/2006 5:04",
		"1/02/2006 15:04",
		"1/02/2006 5:04",
		"1/2/2006 15:04",
		"1/2/2006 5:04",
	}
	for _, layout := range layouts {
		if date, err := time.Parse(layout, dateString); err == nil {
			return date, nil
		}
	}
	return time.Time{}, fmt.Errorf("invalid date format: %s", dateString)
}
